bubblegum/xrf/model/fit_spectrum.py

```diff
@@ -0,0 +1,226 @@
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES   #
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR   #
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)   #
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,  #
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OTHERWISE) ARISING   #
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   #
+# POSSIBILITY OF SUCH DAMAGE.                                          #
+########################################################################
+
+__author__ = 'Li Li'
+
+import numpy as np
+import time
+import json
+import six
+import os
+
+from atom.api import Atom, Str, observe, Typed, Int, List, Dict, Float
+
+from skxray.fitting.xrf_model import (k_line, l_line, m_line)
+from skxray.constants.api import XrfElement as Element
+
+from skxray.fitting.xrf_model import (ModelSpectrum, update_parameter_dict,
+                                      get_sum_area, set_parameter_bound,
+                                      ParamController, set_range, get_linear_model,
+                                      PreFitAnalysis, k_line, l_line)
+from skxray.fitting.background import snip_method
+from lmfit import fit_report
+
+out_folder = '/Users/Li/Research/X-ray/Research_work/all_code/nsls2_gui/nsls2_gui'
+
+class Fit1D(Atom):
+
+    file_path = Str()
+    file_status = Str()
+    param_dict = Dict()
+    data = Typed(np.ndarray)
+    fit_x = Typed(np.ndarray)
+    fit_y = Typed(np.ndarray)
+    residual = Typed(np.ndarray)
+    comps = Dict()
+    fit_strategy1 = Int(0)
+    fit_strategy2 = Int(0)
+    fit_strategy3 = Int(0)
+    fit_result = Typed(object)
+    strategy_list = List()
+    data_title = Str()
+    result_folder = Str(out_folder)
+
+    def __init__(self):
+        self.strategy_list = ['fit_with_tail', 'free_more', 'e_calibration', 'linear']
+
+    def get_full_param(self):
+        try:
+            with open(self.file_path, 'r') as json_data:
+                self.param_dict = json.load(json_data)
+            self.file_status = 'Parameter file {} is loaded.'.format(self.file_path.split('/')[-1])
+        except ValueError:
+            self.file_status = 'Parameter file can\'t be loaded.'
+
+    @observe('file_path')
+    def update_param(self, change):
+        self.get_full_param()
+
+    @observe('fit_strategy1')
+    def update_strategy1(self, change):
+        if change['value'] == 0:
+            return
+        set_parameter_bound(self.param_dict,
+                            self.strategy_list[self.fit_strategy1-1])
+
+    @observe('fit_strategy2')
+    def update_strategy2(self, change):
+        print('strateg2 changed')
+
+    @observe('fit_strategy3')
+    def update_strategy3(self, change):
+        print('strateg3 changed')
+
+    def update_param_with_result(self):
+        update_parameter_dict(self.param_dict, self.fit_result)
+
+    def fit_data(self):
```

Comment:
We can think about that later. Right now I only put the most fundamental fitting functions in skxray.